src/binfileutils.js

```diff
@@ -10,62 +14,143 @@ export async function readBinFile(fileName, type, maxVersion, cacheSize, pageSiz
     let readedType = "";
     for (let i=0; i<4; i++) readedType += String.fromCharCode(b[i]);
 
-    if (readedType != type) throw new Error(fileName + ": Invalid File format");
+    if (readedType !== type) {
+        fd.close();
+        throw new Error(fileName + ": Invalid File format");
+    }
 
     let v = await fd.readULE32();
 
-    if (v>maxVersion) throw new Error("Version not supported");
+    let [version, binVersion] = decodeVersion(v);
 
-    const nSections = await fd.readULE32();
+    if (version > maxVersion) {
+        fd.close();
+        throw new Error("Version not supported");
+    }
 
-    // Scan sections
     let sections = [];
-    for (let i=0; i<nSections; i++) {
-        let ht = await fd.readULE32();
-        let hl = await fd.readULE64();
-        if (typeof sections[ht] == "undefined") sections[ht] = [];
-        sections[ht].push({
-            p: fd.pos,
-            size: hl
-        });
-        fd.pos += hl;
+    if(BIN_FORMAT_1 === binVersion) {
+        await readBinFileV1();
+    } else {
+        await readBinFileV2();
     }
 
+    fd.binVersion = binVersion;
+
     return {fd, sections};
+
+    async function readBinFileV1() {
+        const nSections = await fd.readULE32();
+
+        // Scan sections
+        for (let i = 0; i < nSections; i++) {
+            let ht = await fd.readULE32();
+            let hl = await fd.readULE64();
+            if (typeof sections[ht] == "undefined") sections[ht] = [];
+            sections[ht].push({
+                p: fd.pos,
+                size: hl
+            });
+            fd.pos += hl;
+        }
+    }
+
+    async function readBinFileV2() {
```

Comment:
Do you want these functions inside the closure instead of passing in arguments you need?